File: sgp.R
#!/usr/bin/env -S Rscript --vanilla

# SuperGenPass in R. Made by mnl@vandal.nu 2018
# Make sure to: package.install("openssl")

library("openssl")

# Get and handle arguments
args <- commandArgs(trailingOnly=TRUE)
if (length(args)==0) { stop("Usage ./sgp.R domainname [length] [digest]") }
domain <- args[1]
dig = switch ( args[2],	
							sha256 = {sha256}, 
							sha512 = {sha512}, 
							{md5} ) # Default digest and for unknown entries
len <- ifelse(length(args) > 2,args[3],15) # Default hash length

# Get secure input from R
if (interactive()) { pass <- askpass("Password: ") 
# Rscript is considered non-interactive so a workaround is needed:
} else { cat("Password: "); pass <- readLines("stdin",n=1);} 

hash <- paste(pass,domain,sep=':')

reapply <- function(hash){
	# Do hashing and translate to alphanumerical characters
	hash <- base64_encode(dig(charToRaw(hash)))
	hash <- chartr('+/=','98A',hash)
	return(hash)
}
valid <- function(hash){
	# Check hash for starting lowercase, digits and uppercase
	if ((substring(hash,0,1) %in% letters) &&
			(!identical(hash,tolower(hash))) &&
			(grepl('[0-9]',hash)))
	{
		return(TRUE)
	}
	return(FALSE)
} # Loop 10 times and the until hash is valid SGP Pass
for (i in 1:1000) {
	hash <- reapply(hash)
	if (i <= 10) { next }
	if (valid(hash)) { break }
}
# Done! Cut to length and print
hash = substring(hash,0,min(len,nchar(hash)))
cat(paste(hash,"\n"))

File: sgp.jl
#!/usr/bin/env julia
#=
SuperGenPass version in Julia. Made in 2018 by mnl@vandal.nu
Base64 and SHA are in stdlib. MD5 or Nettle needs to be Pkg.add:ed
=#

using MD5, Base64 #Pkg.add("MD5") 

len = 15 # Set defaults
dig = md5
if (length(ARGS) < 1) # Get argv
	print(stderr, "Usage "*PROGRAM_FILE*" [domainname] [length, digest (optional)]\n")
  exit(1)
elseif (length(ARGS) > 1)
	len = tryparse(Int, ARGS[2])
	if (length(ARGS) > 2)
		if (occursin("sha",ARGS[3]))
			using SHA
			dig = sha256
		end
	end
end

# Get input
pass = Base.getpass("Password")
hash = read(pass,String) * ":" * ARGS[1]
Base.shred!(pass)

translate(c) = get(Dict('+' => '9', '/' => '8', '=' => 'A'),c,c) # poor mans tr!

function sgp!(h)
	return map(translate, base64encode(dig(h)))
end
function sgp!(n::Int) # Multiple dispatch, yeah
	if (n>1)
		global hash = sgp!(hash)
		return sgp!(n-1)
	end
	return sgp!(hash)
end

function valid(hash) 
	if (islowercase(hash[1]) && # Testing out native Julia functions instead of regex
			findfirst(isuppercase, hash) != nothing && 
			findfirst(isdigit, hash) != nothing)
      return true
	end
  return false
end

hash = sgp!(10)
while(!valid(hash))
	global hash = sgp!(hash)
end

println(hash[1:min(len,end)]) # Print and cut to lenght 
